b2/raw_api.py

```diff
@@ -0,0 +1,458 @@
+        """
+        for k in headers:
+            if k.lower() == 'user-agent':
+                return headers
+        else:
+            result = dict(headers)
+            result['User-Agent'] = USER_AGENT
+            return result
+
+    def __enter__(self):
+        try:
+            request = urllib.request.Request(self.url, self.data, self.headers)
+            self.file = urllib.request.urlopen(request)
+            return self.file
+        except urllib.error.HTTPError as e:
+            data = e.read()
+            raise WrappedHttpError(data, self.url, self.params, self.headers, sys.exc_info())
+        except urllib.error.URLError as e:
+            raise WrappedUrlError(e.reason, self.url, self.params, self.headers, sys.exc_info())
+        except socket.error as e:  # includes socket.timeout
+            # reportedly socket errors are not wrapped in urllib2.URLError since Python 2.7
+            raise WrappedSocketError(
+                'errno=%s' % (e.errno,),
+                self.url,
+                self.params,
+                self.headers,
+                sys.exc_info(),
+            )
+        except six.moves.http_client.HTTPException as e:  # includes httplib.BadStatusLine
+            raise WrappedHttplibError(str(e), self.url, self.params, self.headers, sys.exc_info())
+
+    def __exit__(self, exception_type, exception, traceback):
+        if self.file is not None:
+            self.file.close()
+
+
+def read_str_from_http_response(response):
+    """
+    This is an ugly hack.  I probably don't understand Python 2/3
+    compatibility well enough.
+
+    The read() method on urllib responses returns a str in Python 2,
+    and bytes in Python 3, so json.load() won't work in both. This
+    function converts the result into a str that json.loads() will
+    take.
+    """
+    if six.PY2:
+        return response.read()
+    else:
+        return response.read().decode('utf-8')
+
+
+def post_json(url, params, auth_token=None):
+    """Coverts params to JSON and posts them to the given URL.
+
+    Returns the resulting JSON, decoded into a dict or an
+    exception, custom if possible, WrappedHttpError otherwise
+    """
+    data = six.b(json.dumps(params))
+    headers = {}
+    if auth_token is not None:
+        headers['Authorization'] = auth_token
+    try:
+        with OpenUrl(url, data, headers, params) as f:
+            json_text = read_str_from_http_response(f)
+            return json.loads(json_text)
+    except WrappedHttpError as e:
+        e_backup = sys.exc_info()
+        try:
+            error_dict = json.loads(e.data.decode('utf-8'))
+            raise map_error_dict_to_exception(e, error_dict, params)
+        except ValueError:
+            v = sys.exc_info()
+            error_dict = {'error_decoding_json': v}
+            raise FatalError('error decoding JSON when handling an exception', [e_backup, v],)
+
+
+@six.add_metaclass(ABCMeta)
+class RawApi(object):
```

Comment:
I suggest to rename to `AbstractRawApi` for consistency with other abstract classes that we have